Update psa-thread-safety.md to reflect version 3.6 changes (#8928)
Mbed-TLS/mbedtls: docs/architecture/psa-thread-safety/psa-thread-safety.md

- The testing system has now been made thread-safe. Tests can now spin up multiple threads, see [Thread-safe testing](#thread-safe-testing) for details.
- Some multithreaded testing of the key management API has been added, this is outlined in [Testing-and-analysis](#testing-and-analysis).
- The solution uses the pre-existing `MBEDTLS_THREADING_C` threading abstraction.
- The core makes no additional guarantees for drivers. See [Driver policy](#driver-policy) for details.

The usage of keys within other PSA Crypto APIs is planned to be made thread-safe in future, but currently we are not testing this.

## Overview of the document

* The [Guarantees](#guarantees) section describes the properties that are followed when PSA functions are invoked by multiple threads.
* The [Usage guide](#usage-guide) section gives guidance on initializing, using and freeing PSA when using multiple threads.
* The [Current strategy](#current-strategy) section describes how thread-safety of key management and `global_data` is achieved.
* The [Testing and analysis](#testing-and-analysis) section discusses the state of our testing, as well as how this testing will be extended in future.

> * The [Future work](#future-work) section outlines our long-term goals for thread-safety; it also analyses how we might go about achieving these goals.

Reviewer: Typo (I think):
```suggestion
* The [Future work](#future-work) section outlines our long-term goals for thread-safety; it also analyzes how we might go about achieving these goals.
```

Author: Analyses is British English and analyzes is American - either is correct, I have tried to just use British English for this document.

Reviewer: Ops, sorry for the wrong comment then :sweat_smile:

## Guarantees

### Correctness out of the box

Building with `MBEDTLS_PSA_CRYPTO_C` and `MBEDTLS_THREADING_C` gives code which is correct; there are no race-conditions, deadlocks or livelocks when concurrently calling any set of PSA key management functions once `psa_crypto_init` has been called (see the [Initialization](#initialization) section for details on how to correctly initialize the PSA subsystem when using multiple threads).

We do not test or support calling other PSA API functions concurrently.

There is no busy-waiting in our implementation, every API call completes in a finite number of steps regardless of the locking policy of the underlying mutexes.

When only considering key management functions: Mbed TLS 3.6 abides by the minimum expectation for concurrent calls set by the PSA specification (see [PSA Concurrent calling conventions](#psa-concurrent-calling-conventions)).

#### PSA Concurrent calling conventions

These are the conventions which are planned to be added to the PSA 1.2 specification, Mbed TLS 3.6 abides by these when only considering [key management functions](https://arm-software.github.io/psa-api/crypto/1.1/api/keys/management.html):

> > The result of two or more concurrent calls must be consistent with the same set of calls being executed sequentially in some order, provided that the calls obey the following constraints:

Author: What I am quoting hasn't yet been uploaded to the PSA spec, it is from https://github.com/ARM-software/psa-api/pull/142. It is planned to be added to PSA 1.2. The current version of the spec doesn't allow you to return resource error codes in the way we return them

##### Re-loading persistent keys

As described above, persistent keys can be kicked out of the key slot array provided they are not currently being used (`registered_readers == 0`). When attempting to use a persistent key that has been kicked out of a slot, the call to `psa_get_and_lock_key_slot` will see that the key is not in a slot, call `psa_reserve_free_key_slot` and load the key back into the reserved slot. This entire sequence is done during a single mutex lock, which is necessary for thread-safety (see documentation of `psa_get_and_lock_key_slot`).

If `psa_reserve_free_key_slot` cannot find a suitable slot, the key cannot be loaded back in. This will lead to a `PSA_ERROR_INSUFFICIENT_MEMORY` error.

##### Using existing keys

One-shot operations follow a standard pattern when using an existing key:

* They call some `psa_get_and_lock_key_slot_X` function, which finds the key and registers the thread as a reader.
* They operate on the key slot, usually copying the key into a separate buffer to be used by the operation. This step is not done under the mutex.
* Once finished, they call `psa_unregister_read_under_mutex`.

> Multi-part and restartable operations each have a "setup" function where the key is inputted. This function follows the above pattern. The key is copied into the `operation` object, and the thread unregisters. They do not access the key slots again. The copy of the key will not be destroyed during a call to `psa_destroy_key`, the thread running the operation is responsible for deleting this copy in the clean-up. This may need to change to enforce the long term key requirements ([Long term key destruction requirements](#long-term-key-destruction-requirements)).

Reviewer 2: ```suggestion
Multi-part and restartable operations each have a "setup" function where the key is passed in, and this function then follows the above pattern. The key is copied into the `operation` object, and the thread unregisters its reader (they do not access the key slots again). The copy of the key will not be destroyed during a call to `psa_destroy_key`, the thread running the operation is responsible for deleting its copy of the key in the clean-up. This may need to change to enforce the long term key requirements ([Long term key destruction requirements](#long-term-key-destruction-requirements)).
```

##### Key destruction implementation

The locking strategy here is explained in `library/psa_crypto.c`. The destroying thread (the thread calling `psa_destroy_key`) does not always wipe the key slot. The destroying thread registers to read the key, sets the slot's state to `PSA_SLOT_PENDING_DELETION`, wipes the slot from memory if the key is persistent, and then unregisters from reading the slot.

> `psa_unregister_read` internally calls `psa_wipe_key_slot` iff the slot's state is `PSA_SLOT_PENDING_DELETION` and the slot's registered reader counter is equal to 1. This implements a "last one out closes the door" approach, where the final thread to unregister from reading a destroyed key will automatically wipe the contents of the slot; this ensure that there is no corruption.

Reviewer 2: ```suggestion
`psa_unregister_read` internally calls `psa_wipe_key_slot` if the slot's state is `PSA_SLOT_PENDING_DELETION` and the slot's registered reader counter is equal to 1. This implements a "last one out closes the door" approach, where the final thread to unregister from reading a destroyed key will automatically wipe the contents of the slot; this will ensure that there is no corruption as no readers are left to reference the slot post deletion.
```

Author: The "iff" here is purposeful - short hand for "if and only if". I'll switch to using the actual words as this is a point of confusion.

Reviewer 2: Doh, yeah, should have recognised that, but I think longhand is better.

### Linearisability of the system

To satisfy the requirements in [Correctness out of the box](#correctness-out-of-the-box), we require our functions to be "linearisable" (under certain constraints). This means that any (constraint satisfying) set of concurrent calls are performed as if they were executed in some sequential order.

> The standard way of reasoning that this is the case is to identify a "linearization point" for each call, this is a single execution step where the function takes effect (this is usually a step in which the effects of the call become visible to other threads). If every call has a linearization point, the set of calls is equivalent to sequentially performing the calls in order of when their linearization point occurred.

Reviewer 2: ```suggestion
The standard way of reasoning that this is the case is to identify a "linearisation point" for each call, this is a single execution step where the function takes effect (this is usually a step in which the effects of the call become visible to other threads). If every call has a linearization point, the set of calls is equivalent to sequentially performing the calls in order of when their linearization point occurred.
```
S or Z? I don't particularly care, but we should probably pick one..

To help justify that our system is linearisable, here are the linearization points/planned linearization points of each PSA call :

* Key creation functions (including `psa_copy_key`) - The linearization point for a successful call is the mutex unlock within `psa_finish_key_creation`; it is at this point that the key becomes visible to other threads. The linearization point for a failed call is the closest mutex unlock after the failure is first identified.
* `psa_destroy_key` - The linearization point for a successful destruction is the mutex unlock, the slot is now in the state `PSA_SLOT_PENDING_DELETION` meaning that the key has been destroyed. For failures, the linearization point is the same.
* `psa_purge_key`, `psa_close_key` - The linearization point is the mutex unlock after wiping the slot for a success, or unregistering for a failure.

> * One shot operations - The linearization point is the final unlock of the mutex within `psa_get_and_lock_key_slot`, as that is the point in which it is decided whether the key exists.

Reviewer 2: ```suggestion
* One shot operations - The linearization point is the final unlock of the mutex within `psa_get_and_lock_key_slot`, as that is the point in which it is decided whether or not the key exists.
```
Hamlet.

* Multi-part operations - The linearization point of the key input function is the final unlock of the mutex within `psa_get_and_lock_key_slot`. All other steps have no non resource-related side effects (except for key derivation, covered in the key creation functions).

> Please note that one shot operations and multi-part operations are not yet considered thread-safe, as we do not test whether they rely on unprotected global resources.

Reviewer 2: ```suggestion
Please note that one shot operations and multi-part operations are not yet considered fully thread-safe, as we have not yet tested whether they rely on unprotected global resources. Only the key slot access parts have been made safe.
```

## Testing and analysis

### Thread-safe testing

It is now possible for individual tests to spin up multiple threads. This work has made the global variables used in tests thread-safe. If multiple threads fail a test assert, the first failure will be reported with correct line numbers.

> The `step` feature used in some tests is not thread-safe, it cannot be made thread-safe unless we introduce thread-local variables.

Reviewer 2: ```suggestion
It should be noted that although the `step` feature used in some tests is thread-safe, it may produce unexpected results. `mbedtls_test_set_step` or `mbedtls_test_increment_step` calls within threads may happen in any order, thus may not produce the desired result if precise ordering was required.
```

### Current state of testing

> Our testing is a work in progress. It is not feasible to run our traditional, single-threaded, tests in such a way that tests concurrency. Therefore, we must write a new suite of testing.

Reviewer 2: ```suggestion
Our testing is a work in progress. It is not feasible to run our traditional, single-threaded, tests in such a way that tests concurrency, therefore, we need to write new test suites to do this.
```

> Our tests currently only run on pthread.

Reviewer 2: ```suggestion
Our tests currently only run on pthread, this is something we hope to expand in the future (Once again we have an API that should allow this)
```

> We run tests using [ThreadSanitizer](https://clang.llvm.org/docs/ThreadSanitizer.html) to detect data races. We test the key store, and test that our key slot state system is enforced.

Reviewer 2: Should we mention the init tests here?

> Currently, not every API call is tested, we also cannot feasibly test every combination of concurrent API calls. API calls can in general be split into a few categories, each category calling the same internal functions in the same order - it is the internal functions that are in charge of locking mutexes and interacting with the key store; we have tests which cover every category.

Reviewer 2: ```suggestion
Currently, not every API call is tested, we also cannot feasibly test every combination of concurrent API calls. API calls can in general be split into a few categories, each category calling the same internal functions in the same order - it is the internal functions that are in charge of locking mutexes and interacting with the key store, and fot these we have tests which cover every category.
```
Not entirely sure I'm getting the gist of this correctly.

> Since we do not run every cryptographic operation concurrently, we do not test that operations are free of unexpected global variables, cryptographic operations are not considered thread-safe.

Reviewer 2: We have kind of said this before. do we need to repeat it?

Author: Good point: I'll say that we don't test for globals as that is what this section is about, but I'll get rid of the last clause.

### Expanding testing

Through future work on testing, it would be good to:

* For every API call, have a test which runs multiple copies of the call simultaneously.
* After implementing other threading platforms, expand the tests to these platforms.
* Have increased testing for kicking persistent keys out of slots.

> * Explicitly test that there all global variables are protected, for this we need to cover every operation in a concurrent scenario while running ThreadSanitizer.

Reviewer 2: ```suggestion
* Explicitly test that all global variables are protected, for this we would need to cover every operation in a concurrent scenario while running ThreadSanitizer.
```

### Performance

Key loading does somewhat run in parallel, deriving the key and copying it key into the slot is not done under any mutex.

Key destruction is entirely sequential, this is required for persistent keys to stop issues with re-loading keys which cannot otherwise be avoided without changing our approach to thread-safety.

## Future work

### Long term requirements

As explained previously, we eventually aim to make the entirety of the PSA API thread-safe. This will build on the work that we have already completed. This requires a full suite of testing, see [Expanding testing](#expanding-testing) for details.

> ### Long term performance requirements

Reviewer 2: We should mention our desire to replace some (if not all) mutexes with RWLocks if at all possible.

> Our plan for cryptographic operations is that they are not performed under any global mutex. One-shot operations and multi-part operations will each only hold the global mutex for finding the relevant key in the key slot, and unregistering as a reader after the operation.

Reviewer 2: ```suggestion
Our plan for cryptographic operations is that they are not performed under any global mutex. One-shot operations and multi-part operations will each only hold the global mutex for finding the relevant key in the key slot, and unregistering as a reader after the operation, using their own mutexes to guard any shared data that they use.
```

Here are our long term key destruction goals:

`psa_destroy_key` does not block indefinitely, and when `psa_destroy_key` returns:

1. The key identifier does not exist. This is a functional requirement for persistent keys: any thread can immediately create a new key with the same identifier.
2. The resources from the key have been freed. This allows threads to create similar keys immediately after destruction, regardless of resources.
4. No copy of the key material exists. Rationale: this is a security requirement. We do not have this requirement yet, but we need to document this as a security weakness, and we would like to satisfy this security requirement in the future.

#### Condition variables

It would be ideal to add these to a future major version; we cannot add these as requirements to the default `MBEDTLS_THREADING_C` for backwards compatibility reasons.

Condition variables would enable us to fulfil the final requirement in [Long term key destruction requirements](#long-term-key-destruction-requirements). Destruction would then work as follows:

 * When a thread calls `psa_destroy_key`, they continue as normal until the `psa_unregister_read` call.

>  * Instead of calling `psa_unregister_read`, the thread waits until the condition `slot->registered_readers == 1` is true (the destroying thread is the final reader).

Reviewer 2: Isn't this basically a forbidden spin lock though?

Author: I may be wrong here but isn't it the case that condition variables are not spin locking, as one thread sleeps until another thread sends a wake-up signal?

Reviewer 2: Ok, after discussion, I am happy to cede this point. Spin locking without sleeping is much more evil than sleeping the thread to wait, but does cause some of the same issues with timing.

* By default, each driver only has at most one entry point active at any given time. In other words, each driver has its own exclusive lock.
* Drivers have an optional `"thread_safe"` boolean property. If true, it allows concurrent calls to this driver.
* Even with a thread-safe driver, the core never starts the destruction of a key while there are operations in progress on it, and never performs concurrent calls on the same multipart operation.

The purpose of `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` is very similar to the driver interface (and might even be added to it in the long run), therefore it makes sense to handle it the same way. In particular, we can use the `global_data` mutex to protect it as a default and when we implement the "thread_safe” property for drivers, we implement it for `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` as well.
In the non-thread-safe case we have these natural assumptions/requirements:

#### Implementation notes
1. Drivers don't call the core for any operation for which they provide an entry point.
2. The core doesn't hold the driver mutex between calls to entry points.

> Since we only have simple mutexes, locking the same mutex from the same thread is a deadlock. Therefore functions taking the global mutex must not be called while holding the same mutex. Functions taking the mutex will document this fact and the implications.

Reviewer 2: Not sure we should have lost these two implementation notes. Not entirely sure where they fit, however.

Reviewer 2: Missed one, raised another review for this.
